units/helpers.go

```diff
@@ -0,0 +1,22 @@
+package units
+
+import (
+	"context"
+	"io"
+)
+
+type TimeoutReadCloser struct {
+	ctx context.Context
+	src io.ReadCloser
+}
+
+func (r *TimeoutReadCloser) Read(p []byte) (n int, err error) {
+	if err := r.ctx.Err(); err != nil {
```

Comment:
I don't understand how it works.
 IMO if context is cancelled but we can read and finish operation we should do it. If reading takes long and context could be cancelled in meantime we should cancel it. This only works when context is cancelled before reading which is not something we are looking for.

Reply:
Let me elaborate on how this works:

* A context is created at time 0 with deadline 5.
* At time 1 `Read` is called and returns the desired amount of bytes (at most).
* At time 2 `Read` is called and returns the desired amount of bytes (at most).
* At time 5 the deadline is reached.
* At time 6 `Read` is called but since the deadline has already been reached we don't read anything but return immediately.

What this does NOT account for is that when during a `Read` operation the deadline is reached. The `Read` operation of the underlying source will NOT be interrupted. But this is IMO an edge case we can probably live with. It's more important to read as much data as possible in the given time frame and stop reading when time's up.

Reply:
OK, got it!

Reply:
Related: https://github.com/golang/go/issues/20280